To make subfolder's exports visible.

zjkuang opened the pull request:
<!-- Please provide enough information so that others can review your pull request. -->
<!-- Keep pull requests small and focused on a single change. -->

### Summary

The original template put all things in `src/index.tsx` and exported everything from there. As the package grows, we might want to organize things into subfolders. e.g.
```
/src/native/index.ts
/src/utils/index.ts
/src/index.tsx
```
and in `/src/index.tsx` we have
```
export * from './native';
export * from './utils';
```
so that all the members exported from `/src/native/index.ts` and `/src/utils/index.ts` would finally be exported through `/src/index.tsx` and should be accessible by this package's consumer project.

There is no problem doing so when this package is "local". (e.g. when we run `/example` provided along with the package by the template, this package is "local".) But if an outside project installed this package from `npmjs` (or `yarnpkg`) and `import {...} from '<this-package>';`, all the exported members became invisible.

To fix this issue, replace `"lib"` with `"lib/**/*"` in `package.json`'s `"files"` entry.

### Test plan

<!-- List the steps with which we can test this change. Provide screenshots if this changes anything visual. -->

Verify this fix as described in the above `Summary` section.

vonovak commented:
Hello @zjkuang and thanks for opening the PR!
I do not believe the change you're proposing is necessary: `lib` and `lib/**/*` are equivalent.

E.g. for the `android` folder, we also specify just `android` - no package built using `react-native-builder-bob` would work if this was wrong (the android folder has multiple levels of folder nesting).

If you're sure that this is necessary, can you add some more evidence?

thank you :)

satya164 commented:
I agree with @vonovak. The entries in `files` only control what gets published to npm, they can just be directory names and not globs. They don't have any effect on what you can import.

In addition, I'd also add that imports like `import something from 'library/lib/commonjs/something` isn't advisable. There are many issues with this:

1. There are different targets for different environments (`commonjs` for Node, `esm` for Webpack, etc.), importing it this way doesn't work.
2. For the majority of the packages, these files are internal. Users importing them this way means any small changes to the file structure are a breaking change, this sort of usage should be explicitly supported by the library developer if they want to. So even if there was a restriction, it should be an explicit decision by the maintainer and not supported by default.

You don't have to export everything like in your example, exports from various modules could be grouped under an object and exported as well.

If you really want to have folders instead, you should create folders in the root of the project and create a package.json with `main`, `module` and `react-native` entries that points to the relevant files under `lib` so the user can import like `import something from 'library/something'`. But this is not something that belongs in the default template. So I'm closing this PR.

zjkuang commented:
Hi @vonovak & @satya164

Thanks for the quick response. Honestly, it's my first time trying to publish a react-native package in TypeScript and your builder-bob is really amazing! As for the issue I met, I published my test package `@zjkuang/react-native-utils`. I was using `yarn add @zjkuang/react-native-utils` to install this test package to my other app. I found that if in the package I moved the functions+exports to other places than /src/index.ts, the consumer app couldn't see the functions anymore. I didn't know what was wrong and tried many things with no luck until I replaced "lib" with "lib/**/*" and then everything was working like a charm. I changed "lib/**/*" back to "lib" and the issue came back again. So this issue is reproducible. I agree that "lib" "lib/**/*" should be equivalent but so far I couldn't find any other right way to fix it than this change. (And I thought other users might also run into the same situation so I created this PR.) Both my package and GitHub repo are public. If you could take a look and help me find what is wrong in my package and how to fix it in the right way (if "lib/**/*" doesn't make sense), it would be great.

satya164 commented:
@zjkuang The `files` field is used by npm to package and publish your library, you can see what's included using `npm pack` command. As per my knowledge, neither metro (the bundler used for React Native) nor Webpack (common on Web) look at the `files` field to determine what can be imported. I have imported files like this in many projects and I double-checked again with a library that doesn't have a glob in `files` to be sure.

I'm not sure what happened in your case, or what you mean by the consumer app couldn't see the files. Do you mean it couldn't see the types? Did import fail with metro, webpack, or any other bundler? Are you using something like Yarn PnP or PNPM? Though most now respect the `exports` field to restrict what can be imported which we don't currently use.

Regardless, even if this doesn't work in a bundler, it's not something that should be supported by default due to the reasons I mentioned. Making internal folders part of the public API should be an explicit decision by the library maintainer.

zjkuang commented:
> @zjkuang The `files` field is used by npm to package and publish your library, you can see what's included using `npm pack` command. As per my knowledge, neither metro (the bundler used for React Native) nor Webpack (common on Web) look at the `files` field to determine what can be imported. I have imported files like this in many projects and I double-checked again with a library that doesn't have a glob in `files` to be sure.
> 
> I'm not sure what happened in your case, or what you mean by the consumer app couldn't see the files. Do you mean it couldn't see the types? Did import fail with metro, webpack, or any other bundler? Are you using something like Yarn PnP or PNPM? Though most now respect the `exports` field to restrict what can be imported which we don't currently use.
> 
> Regardless, even if this doesn't work in a bundler, it's not something that should be supported by default due to the reasons I mentioned. Making internal folders part of the public API should be an explicit decision by the library maintainer.

Sorry about not making it clear. By saying "the consumer app couldn't see the exported members", we could see it from the VSCode:
<img width="879" alt="Screen Shot 2022-08-28 at 6 28 59 PM" src="https://user-images.githubusercontent.com/46003022/187229792-39f9d0a3-f066-4641-a75d-77589f21dec5.png">
